twilight-cache-inmemory/src/traits.rs

```diff
@@ -0,0 +1,247 @@
+
+    /// Update the cached data with a [`GuildUpdate`] event. Fields containing other
+    /// cached structures such as channels are cleared prior.
+    fn update_with_guild_update(&mut self, guild_update: &GuildUpdate);
+
+    /// Increase the guild member count.
+    fn increase_member_count(&mut self, amount: u64);
+
+    /// Decrease the guild member count.
+    fn decrease_member_count(&mut self, amount: u64);
+}
+
+/// Trait for a generic cached representation of a [`VoiceState`].
+pub trait CacheableVoiceState:
+    From<(Id<ChannelMarker>, Id<GuildMarker>, VoiceState)> + Debug
+{
+    /// ID of the channel this voice state belongs to.
+    fn channel_id(&self) -> Id<ChannelMarker>;
+}
+
+/// Trait for a generic cached representation of a [`Message`].
+pub trait CacheableMessage: From<Message> + Debug {
+    /// Update the cached data with a [`MessageUpdate`] event.
+    fn update_with_message_update(&mut self, message_update: &MessageUpdate);
+
+    /// Reactions added to this message.
+    fn reactions(&self) -> &[Reaction];
+
+    /// Mutable getter for reactions added to this message.
+    fn reactions_mut(&mut self) -> &mut [Reaction];
+
+    /// Retain all reactions to this message matching a predicate, removing non-matching ones.
+    fn retain_reactions(&mut self, f: impl FnMut(&Reaction) -> bool);
+
+    /// Clear all reactions to this message.
+    fn clear_reactions(&mut self);
+
+    /// Add a reaction to this message.
+    fn add_reaction(&mut self, reaction: Reaction);
+
+    /// Remove a reaction from this message.
+    fn remove_reaction(&mut self, idx: usize);
+}
+
+/// Trait for a generic cached representation of a [`CurrentUser`].
+pub trait CacheableCurrentUser: From<CurrentUser> + Clone + Debug {
+    /// ID of the user.
+    fn id(&self) -> Id<UserMarker>;
+}
+
+impl CacheableCurrentUser for CurrentUser {
+    fn id(&self) -> Id<UserMarker> {
+        self.id
+    }
+}
+
+/// Trait for a generic cached representation of a [`Sticker`].
+pub trait CacheableSticker: From<Sticker> + PartialEq<Sticker> + Debug {
+    /// ID of the sticker.
+    fn id(&self) -> Id<StickerMarker>;
+}
+
+/// Trait for a generic cached representation of a [`Emoji`].
+pub trait CacheableEmoji: From<Emoji> + PartialEq<Emoji> + Debug {}
+
+/// Trait for a generic cached representation of a [`GuildIntegration`].
+pub trait CacheableGuildIntegration: From<GuildIntegration> + PartialEq<Self> + Debug {}
+
+impl CacheableGuildIntegration for GuildIntegration {}
+
+/// Trait for a generic cached representation of a [`Presence`].
+pub trait CacheablePresence: From<Presence> + Debug {}
+
+/// Trait for a generic cached representation of a [`StageInstance`].
+pub trait CacheableStageInstance: From<StageInstance> + PartialEq<Self> + Debug {}
+
+impl CacheableStageInstance for StageInstance {}
+
+/// Trait for a generic cached representation of a [`User`].
+pub trait CacheableUser: From<User> + PartialEq<User> + Clone + Debug {}
```

Comment:
We should probably have a base set of required trait implementations for each trait to guard against future proofing. Something like `From<BaseType> + PartialEq<BaseType> + PartialEq<Self> + Clone + Debug`.

Reply:
While I tend agree with `Clone`, `PartialEq` is a problem since the trait requirement does not specify what kind of equality should be checked for and sometimes is not really possible to check for. Let's say a stripped down guild only stores ID and owner ID, but is it really equal to a guild which stores *more* information (aka the full `Guild` struct)? It will not know if a field only present on the other struct has changed in the meantime. Or does the equality check only care about the IDs? It's unclear to the user and poses questions that twilight does not even require the answers to. I would prefer leaving these out and using explicit methods like `.id()` for comparisons, which I think we already use a bit. There are some cases where it is fairly unavoidable, but I think we should use `PartialEq` only where absolutely necessary.

Reply:
Well, `PartialEq` is required for some traits so it seems doable to extend for all traits.

> There are some cases where it is fairly unavoidable, but I think we should use (and require) PartialEq only where absolutely necessary.

"Only where absolutely necessary" is not something we're 100% able to control due to Discord changes, but it's nice to minimize breakage by requiring it just in case.